### PR TITLE
[DEVOPS-53] enable firewall on nodes

### PR DESCRIPTION
**Pull request by deepfire**
Issue: https://issues.serokell.io/issue/DEVOPS-53

### CODE REVIEW

**Comment by deepfire** [comment]
Functionality of a 4-node cluster was tested with firewall enabled:
- nodes find each other and agree on the blockchain tip

**Comment by deepfire** [comment]
Next step:  test cluster functionality with `dhcpcd` disabled.  I have my doubts that this will work, but am ready to be pleasantly surprised : -)

File: modules/cardano-node.nix [hunk 1]
```diff
@@ -142,7 +142,14 @@ in {
       "${toString hour} * * * * root /run/current-system/sw/bin/systemctl restart cardano-node"
     ];
 
-    networking.firewall.allowedTCPPorts = [ cfg.port ];
+    networking.firewall = {
+      enable = mkForce true;
```
Comment (domenkozar) [inline modules/cardano-node.nix]: By default firewall is on. We should get rid of our `false` overrides in the codebase so that this line is redundant.

**Comment by deepfire** [comment]
Disabling `dhcpcd` went smooth, despite my premonitions -- Domen was right. : -)

Next step is refactoring suggested by Domen.

**Comment by deepfire** [comment]
@domenkozar, what would be a test plan for a refactoring with this kind of potential consequences?

**Comment by domenkozar** [comment]
Consequence of my suggestion or this PR in general?

**Comment by deepfire** [comment]
The refactoring you suggest, yes -- it would potentially affect the whole repository -- it shouldn't, of course, if implemented correctly.

The non-refactoring part has been tested, as I have described above.

**Comment by domenkozar** [comment]
We'll have to test as part of next staging deploy that all services work. Really rest of the modules have always considered firewall as being on :)

Meaning, note in this issue that firewall has to be tested on explorer/reportserver as well.

**Comment by deepfire** [comment]
@domenkozar, current situation on the `networking.firewall.enable` value:
- https://issues.serokell.io/_persistent/iohk-nixops-firewall-config.pdf?file=100-469&c=true

From this, I see the following:
- enabled:
   - `cardano-deployer`
   - `cardano-node`
   - `hydra`
- disabled:
   - `cardano-explorer`
   - `report-server`
   - `timewarp`

**Comment by deepfire** [comment]
@domenkozar, pushed the refactoring of `networking.firewall.enable`.

File: modules/timewarp-node.nix [hunk 2]
```diff
@@ -27,7 +27,7 @@ in {
       groups.timewarp = { };
     };
 
-    networking.firewall.allowedTCPPorts = [ cfg.port ];
```
Comment (domenkozar) [inline modules/timewarp-node.nix]: Let's leave the port configured, even if firewall is off.

File: modules/report-server.nix [hunk 3]
```diff
@@ -25,7 +25,7 @@ in {
       groups.report-server = {};
     };
 
-    networking.firewall.allowedTCPPorts = [ cfg.port ];
+    networking.firewall.enabled = false;
```
Comment (domenkozar) [inline modules/report-server.nix]: Actually for report-server, I'd just open up the port. I'm confident it will work.

File: modules/cardano-explorer.nix [hunk 4]
```diff
@@ -14,9 +14,7 @@ with (import ./../lib.nix);
       autoStart = true;
     };
 
-    networking.firewall.allowedTCPPorts = [
```
Comment (domenkozar) [inline modules/cardano-explorer.nix]: We need firewall here enabled as well, let's revert this change. Explorer is also a cardano-node.

Comment (domenkozar) [inline modules/cardano-explorer.nix]: @deepfire removal of these port whitelisting should be reverted

File: modules/report-server.nix [hunk 5]
```diff
@@ -25,8 +25,6 @@ in {
       groups.report-server = {};
     };
 
-    networking.firewall.allowedTCPPorts = [ cfg.port ];
```
Comment (domenkozar) [inline modules/report-server.nix]: @deepfire removal of these port whitelisting should be reverted

Comment (deepfire) [inline modules/report-server.nix]: @domenkozar, done

**Comment by deepfire** [comment]
@domenkozar, there was  a problem testing this on staging: https://issues.serokell.io/issue/DEVOPS-53#comment=96-3955

I'm not really sure how to make it proper -- either we need to patch nginx (which.. well, doesn't seem all bad -- why would an option require the path to exist on nixops evaluator?), or we need to figure out how to make it not to be forced during nixops evaluation

**Comment by deepfire** [comment]
Alright, after investigations, my conclusion was that we cannot disable `dhcpcd` on EC2 instances.  Kind of expected that (https://github.com/input-output-hk/iohk-nixops/pull/51#issuecomment-303846777) -- nodes don't come up.

The surprising part is how this slipped through my testing.

I have a theory -- `nixops modify` isn't as good as a `nixops destroy && nixops delete && nixops create` -- simply because the former doesn't restart the machines.

Hmm. makes sense again : -)

**Comment by deepfire** [comment]
In any case, a freshly-spun 4-node cluster works with the tip of this branch -- and the blockchain tip is agreed upon -- see https://issues.serokell.io/issue/DEVOPS-53#comment=96-4012

@domenkozar, I think this is finally ready.

**Comment by domenkozar** [comment]
@deepfire there are still 2 unaddressed comments

File: modules/timewarp-node.nix (hunk 2 above)
Comment (deepfire) [inline modules/timewarp-node.nix]: Done!

File: modules/cardano-explorer.nix (hunk 4 above)
Comment (deepfire) [inline modules/cardano-explorer.nix]: Done
